Source/Core/Core/HW/WiimoteReal/IOLinux.cpp

```diff
@@ -4,218 +4,315 @@
+  request.length = BLUETOOTH_INQUIRY_LENGTH;
+  request.num_rsp = InquiryRequest::MAX_INFOS;
+  // Use Limited Dedicated Inquiry Access Code (LIAC) like the Wii does.
+  // Third-party Wiimotes cannot be discovered without it.
+  std::ranges::copy(std::to_array({0x00, 0x8b, 0x9e}), request.lap);
+
+  const int hci_inquiry_result = HciInquiry(m_device_sock, &request);
+  switch (hci_inquiry_result)
   {
-    ERROR_LOG_FMT(WIIMOTE, "Error searching for Bluetooth devices.");
+  case 0:
+    break;
+  case ENODEV:
+    Close();
+    [[fallthrough]];
+  default:
+    ERROR_LOG_FMT(WIIMOTE, "Error searching for Bluetooth devices: {}",
+                  Common::StrerrorString(hci_inquiry_result));
     return;
   }
 
-  DEBUG_LOG_FMT(WIIMOTE, "Found {} Bluetooth device(s).", found_devices);
+  DEBUG_LOG_FMT(WIIMOTE, "Found {} Bluetooth device(s).", request.num_rsp);
 
-  // Display discovered devices
-  for (auto& scan_info : scan_infos | std::ranges::views::take(found_devices))
+  for (auto& scan_info : request.scan_infos | std::ranges::views::take(request.num_rsp))
   {
-    // BT names are a maximum of 248 bytes apparently
-    char name[255] = {};
-    if (hci_read_remote_name(m_device_sock, &scan_info.bdaddr, sizeof(name), name, 1000) < 0)
+    const auto bdaddr_str =
+        BluetoothAddressToString(std::bit_cast<Common::BluetoothAddress>(scan_info.bdaddr));
+
+    // Did AddAutoConnectAddresses already add this remote?
+    const auto eq_this_bdaddr = [&](auto* wm) { return wm->GetId() == bdaddr_str; };
+    if (std::ranges::any_of(found_wiimotes, eq_this_bdaddr))
+      continue;
+
+    if (!IsNewWiimote(bdaddr_str))
     {
-      ERROR_LOG_FMT(WIIMOTE, "Bluetooth read remote name failed.");
+      WARN_LOG_FMT(WIIMOTE, "Discovered already connected device: {}", bdaddr_str);
       continue;
     }
 
-    INFO_LOG_FMT(WIIMOTE, "Found bluetooth device with name: {}", name);
+    // The Wii can actually connect remotes with a 10 second name response time.
+    // We won't wait quite so long since we're doing this in a blocking manner right now.
+    // Note that waiting just 1 second can be problematic sometimes.
+    const int read_name_timeout_ms = 3000;
 
-    if (!IsValidDeviceName(name))
+    // BT names are a maximum of 248 bytes.
+    char name[255]{};
+    if (hci_read_remote_name_with_clock_offset(m_device_sock, &scan_info.bdaddr,
+                                               scan_info.pscan_rep_mode, scan_info.clock_offset,
+                                               sizeof(name), name, read_name_timeout_ms) < 0)
+    {
+      ERROR_LOG_FMT(WIIMOTE, "Bluetooth read remote name failed. hci_read_remote_name: {}",
+                    Common::LastStrerrorString());
       continue;
+    }
 
-    char bdaddr_str[18] = {};
-    ba2str(&scan_info.bdaddr, bdaddr_str);
+    INFO_LOG_FMT(WIIMOTE, "Found bluetooth device with name: {}", name);
 
-    if (!IsNewWiimote(bdaddr_str))
+    if (!IsValidDeviceName(name))
```

Comment:
Microoptimization, but both `IsValidDeviceName()` and `IsBalanceBoardName()` construct temporary `std::string`s from this `char[]`, might want to only do this once. Or even better: Let those functions take `std::string_view`.

Reply:
I do change them to `string_view` in #13970

Reply:
Alright then.